standalone/prb/src/pruntime.rs

```diff
@@ -0,0 +1,80 @@
+use anyhow::Result;
+use log::debug;
+use phactory_api::prpc::client::{Error as ClientError, RequestClient};
+use phactory_api::prpc::phactory_api_client::PhactoryApiClient;
+use phactory_api::prpc::server::ProtoError as ServerError;
+use phactory_api::prpc::Message;
+use reqwest::Client;
+use std::future::Future;
+use std::sync::Arc;
+use tokio::sync::Semaphore;
+
+pub type PRuntimeClient = PhactoryApiClient<RpcRequest>;
+
+pub struct RpcRequest {
+    base_url: String,
+    client: Client,
+    semaphore: Arc<Semaphore>,
+}
+
+#[async_trait::async_trait]
+pub trait PRuntimeClientWithSemaphore {
+    async fn with_lock<'a, R>(&'a self, f: (impl Future<Output = R> + Send + 'a)) -> Result<R>;
+}
+
+#[async_trait::async_trait]
+impl PRuntimeClientWithSemaphore for PRuntimeClient {
+    async fn with_lock<'a, R>(&'a self, f: (impl Future<Output = R> + Send + 'a)) -> Result<R> {
+        let s = self.client.semaphore.clone();
+        let s = s.acquire().await?;
+        let ret = f.await;
```

Comment:
Why do we use a Semaphore rather than Mutex?

Reply:
1. It is designed to lock the pruntime requester only when locking needed;
2. The additional `with_lock` method is to achieve 1 without breaking compatibility from functions in pherry crate.

Reply:
If you're meaning that some of the APIs need to be protected, a better approach might be to move those APIs into a separate type called `PRuntimeProtectedAPI` and wrap the new type instance in a Mutex. This could help to prevent mistakes from the caller's side and make the code easier to maintain.

However, keep in mind that neither approach can prevent potential race conditions if there are multiple tasks calling the protected APIs concurrently.

It is not required to refactor it in the current PR, but let's make a note of it in the code comments.